### PR TITLE
Cache setup controls

### PR DESCRIPTION
**Pull request by briankabiro**
#51 

#### What does this PR do?
- add the `last_controls` property to the Client struct
- adds a check before sending the controls to the frontend.

### CODE REVIEW

File: edit-client/src/client.rs [hunk 1]
```diff
@@ -321,13 +322,18 @@ pub trait ClientImpl {
...
+        let mut controls_object = &UserToFrontendCommand::Controls(Controls{
             keys: key_handlers::<Self>()
                 .into_iter()
                 .map(|x| (x.0, x.1, x.2))
                 .collect(),
             buttons: button_handlers::<Self>(state).1
-        })).expect("Could not send initial state");
+        });
+
+        if Some(controls_object) != self.state().last_controls {
```
Comment (briankabiro) [inline edit-client/src/client.rs]: Hey @tcr. Have done most of the work; the only thing left to do now is fixing the error that occurs when comparing the `controls_object` and `last_controls`.

File: edit-client/src/client.rs [hunk 2]
```diff
@@ -321,13 +322,18 @@ pub trait ClientImpl {
     where
         Self: Sized,
     {
-        self.send_client(&UserToFrontendCommand::Controls(Controls{
+        let mut controls_object = &UserToFrontendCommand::Controls(Controls{
```
Comment (tcr) [inline edit-client/src/client.rs]: Here, I think you should make `let controls_object = Controls { ... }` and not wrap it in `UserToFrontendCommand::Controls` yet.

File: edit-client/src/client.rs [hunk 3]
```diff
@@ -321,13 +322,18 @@ pub trait ClientImpl {
...
             keys: key_handlers::<Self>()
                 .into_iter()
                 .map(|x| (x.0, x.1, x.2))
                 .collect(),
             buttons: button_handlers::<Self>(state).1
-        })).expect("Could not send initial state");
+        });
+
+        if Some(controls_object) != self.state().last_controls {
+            self.send_client(controls_object).expect("Could not send initial state");
```
Comment (tcr) [inline edit-client/src/client.rs]: Here you can do `self.send_client(&UserToFrontendCommand::Controls(controls_object.clone()).expect(...)`.

File: edit-client/src/client.rs [hunk 4]
```diff
@@ -321,13 +322,18 @@ pub trait ClientImpl {
...
+            self.state().last_controls = controls_object;
```
Comment (tcr) [inline edit-client/src/client.rs]: I think this needs to be `self.state().last_controls = Some(controls_object);`

**Comment by briankabiro** [comment]
Hey @tcr. No problemo, hope your vacation was great! I'll take a look at the comments; thanks for the pointers.

**Comment by briankabiro** [comment]
Hey @tcr. Thanks for the in-depth comments; helped out a lot. I've made the requested changes. 😃

**Comment by tcr** [comment]
bors try

**Comment by briankabiro** [comment]
Look like the bot timed out. Going to try and run it.

**Comment by briankabiro** [comment]
bors try

**Comment by tcr** [comment]
Hm. Lemme retry:

bors r+

**Comment by tcr** [comment]
bors cancel

**Comment by tcr** [comment]
Okay, I think I fixed everything at last. Let's try again :)

bors r+

**Comment by briankabiro** [comment]
🎆 🎉

**Pull request merged**